File: python-package/mcanalysis/visualization.py
"""
Visualization functions for menstrual cycle analysis.
"""

import pandas as pd
import numpy as np
from typing import List, Optional, Tuple
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches

from .models import GAMResult, PhaseModel


def plot_cycle_effect(
    gam_result: GAMResult,
    inflection_points: Optional[List[float]] = None,
    phase_models: Optional[List[PhaseModel]] = None,
    raw_data: Optional[pd.DataFrame] = None,
    title: str = "Menstrual Cycle Effect on Outcome",
    xlabel: str = "Cycle Day (Day 0 = Period Start)",
    ylabel: str = "% Change from Individual Mean",
    show_significance: bool = True,
    show_phases: bool = True,
    show_linear_models: bool = True,
    show_inflection_points: bool = True,
    show_ci: bool = True,
    show_period_line: bool = True,
    show_raw_data: bool = False,
    day_range: Tuple[int, int] = (-14, 13),
    figsize: Tuple[int, int] = (12, 8),
    gam_color: str = "steelblue",
    gam_linewidth: float = 2.5,
    ci_alpha: float = 0.3,
    raw_data_alpha: float = 0.1,
    raw_data_size: float = 10,
    raw_data_color: str = "gray",
    linear_colors: Optional[List[str]] = None,
    save_path: Optional[str] = None
) -> plt.Figure:
    """
    Plot GAM results with confidence intervals and optional phase information.

    Parameters
    ----------
    gam_result : GAMResult
        Fitted GAM result
    inflection_points : List[float], optional
        Inflection points to mark on plot
    phase_models : List[PhaseModel], optional
        Phase model results to display
    raw_data : pd.DataFrame, optional
        Processed data with 'cycle_day' and 'a' columns for plotting raw observations
    title : str
        Plot title
    xlabel : str
        X-axis label
    ylabel : str
        Y-axis label
    show_significance : bool
        Whether to show significance annotation
    show_phases : bool
        Whether to shade phases differently
    show_linear_models : bool
        Whether to show linear phase model lines (default True)
    show_inflection_points : bool
        Whether to mark inflection points (default True)
    show_ci : bool
        Whether to show confidence interval (default True)
    show_period_line : bool
        Whether to show vertical line at day 0 (default True)
    show_raw_data : bool
        Whether to show raw data points (default False)
    day_range : Tuple[int, int]
        Range of days to display (default -14 to 13)
    figsize : Tuple[int, int]
        Figure size
    gam_color : str
        Color of GAM curve (default "steelblue")
    gam_linewidth : float
        Line width of GAM curve (default 2.5)
    ci_alpha : float
        Transparency of confidence interval (default 0.3)
    raw_data_alpha : float
        Transparency of raw data points (default 0.1)
    raw_data_size : float
        Size of raw data points (default 10)
    raw_data_color : str
        Color of raw data points (default "gray")
    linear_colors : List[str], optional
        Colors for linear phase models (default red and green)
    save_path : str, optional
        Path to save figure

    Returns
    -------
    plt.Figure
        Matplotlib figure object
    """
    if linear_colors is None:
        linear_colors = ["#E74C3C", "#27AE60"]
    pred_df = gam_result.predictions

    # Filter to day range
    mask = (pred_df["cycle_day"] >= day_range[0]) & (pred_df["cycle_day"] <= day_range[1])
    plot_data = pred_df[mask].copy()

    # Convert to percentage change from mean (subtract 100)
    plot_data["predicted_pct"] = plot_data["predicted"] - 100
    plot_data["ci_lower_pct"] = plot_data["ci_lower"] - 100
    plot_data["ci_upper_pct"] = plot_data["ci_upper"] - 100

    fig, ax = plt.subplots(figsize=figsize)

    # Plot daily means as line graph (behind GAM curve)
    if show_raw_data and raw_data is not None:
        # Filter to day range and exclude dummy observations
        raw_mask = (
            (raw_data["cycle_day"] >= day_range[0]) &
            (raw_data["cycle_day"] <= day_range[1]) &
            (~raw_data.get("is_dummy", pd.Series([False] * len(raw_data))).fillna(False))
        )
        raw_plot = raw_data[raw_mask].copy()
        # Convert to percentage change
        raw_plot["a_pct"] = raw_plot["a"] - 100
        # Calculate daily means
        daily_means = raw_plot.groupby("cycle_day")["a_pct"].mean().reset_index()
        daily_means = daily_means.sort_values("cycle_day")

        ax.plot(
            daily_means["cycle_day"],
            daily_means["a_pct"],
            alpha=raw_data_alpha,
            linewidth=raw_data_size / 5,
            color=raw_data_color,
            label="Daily Mean",
            zorder=1
        )

    # Plot confidence interval
    if show_ci:
        ax.fill_between(
            plot_data["cycle_day"],
            plot_data["ci_lower_pct"],
            plot_data["ci_upper_pct"],
            alpha=ci_alpha,
            color=gam_color,
            label="95% CI"
        )

    # Plot GAM curve
    ax.plot(
        plot_data["cycle_day"],
        plot_data["predicted_pct"],
        color=gam_color,
        linewidth=gam_linewidth,
        label="GAM Fit"
    )

    # Add reference line at 0 (mean)
    ax.axhline(y=0, color="gray", linestyle="--", alpha=0.5, label="Mean (0%)")

    # Add vertical line at day 0 (period start)
    if show_period_line:
        ax.axvline(x=0, color="red", linestyle=":", alpha=0.7, label="Period Start")

    # Mark inflection points
    if show_inflection_points and inflection_points:
        for i, ip in enumerate(inflection_points):
            if day_range[0] <= ip <= day_range[1]:
                # Get y value at inflection point (percentage change)
                closest_idx = (plot_data["cycle_day"] - ip).abs().idxmin()
                y_val = plot_data.loc[closest_idx, "predicted_pct"]

                ax.axvline(x=ip, color="orange", linestyle="--", alpha=0.6)
                ax.scatter([ip], [y_val], color="orange", s=100, zorder=5,
                          marker="o", edgecolor="black", linewidth=1)

                # Label
                ax.annotate(
                    f"IP {i+1}\n(Day {ip:.1f})",
                    xy=(ip, y_val),
                    xytext=(10, 10),
                    textcoords="offset points",
                    fontsize=9,
                    bbox=dict(boxstyle="round,pad=0.3", facecolor="white", alpha=0.8)
                )

    # Shade phases if provided
    if show_phases and phase_models and inflection_points:
        colors = plt.cm.Set3(np.linspace(0, 1, len(phase_models)))
        for i, (pm, color) in enumerate(zip(phase_models, colors)):
            if pm.end_day <= pm.start_day:
                # Phase wraps around - shade two regions
                # Region 1: from start_day to day_range[1]
                ax.axvspan(pm.start_day, day_range[1], alpha=0.1, color=color)
                # Region 2: from day_range[0] to end_day
                ax.axvspan(day_range[0], pm.end_day, alpha=0.1, color=color)
            else:
                # Normal phase within range
                start = max(pm.start_day, day_range[0])
                end = min(pm.end_day, day_range[1])
                ax.axvspan(start, end, alpha=0.1, color=color)

    # Add significance annotation
    if show_significance:
        p_val = gam_result.p_value
        sig_text = f"Cycle Effect p-value: {p_val:.2e}"
        if p_val < 0.001:
            sig_text += " ***"
        elif p_val < 0.01:
            sig_text += " **"
        elif p_val < 0.05:
            sig_text += " *"
        else:
            sig_text += " (ns)"

        ax.text(
            0.02, 0.98, sig_text,
            transform=ax.transAxes,
            fontsize=11,
            verticalalignment="top",
            bbox=dict(boxstyle="round,pad=0.5", facecolor="white", edgecolor="gray", alpha=0.9)
        )

        # Add deviance explained
        dev_text = f"Deviance Explained: {gam_result.deviance_explained * 100:.1f}%"
        ax.text(
            0.02, 0.90, dev_text,
            transform=ax.transAxes,
            fontsize=10,
            verticalalignment="top",
            bbox=dict(boxstyle="round,pad=0.5", facecolor="white", edgecolor="gray", alpha=0.9)
        )

    # Add phase model summary and linear model lines if provided
    if phase_models and inflection_points:
        # Get GAM predicted values at turning points (percentage change)
        turning_point_values = {}
        for ip in inflection_points:
            closest_idx = (plot_data["cycle_day"] - ip).abs().idxmin()
            turning_point_values[ip] = plot_data.loc[closest_idx, "predicted_pct"]

        # Plot linear model lines connecting turning points
        if show_linear_models:
            for i, pm in enumerate(phase_models):
                color = linear_colors[i % len(linear_colors)]

                start_y = turning_point_values.get(pm.start_day)
                end_y = turning_point_values.get(pm.end_day)

                if start_y is None or end_y is None:
                    continue

                # Handle cyclical wraparound
                if pm.end_day <= pm.start_day:
                    # Phase wraps around - plot two segments
                    # Segment 1: from start_day to day_range[1]
                    x1_start = pm.start_day
                    x1_end = day_range[1]
                    # Interpolate y values
                    total_span = (day_range[1] - pm.start_day) + (pm.end_day - day_range[0])
                    frac1 = (x1_end - x1_start) / total_span if total_span > 0 else 0
                    y1_end = start_y + frac1 * (end_y - start_y)
                    ax.plot([x1_start, x1_end], [start_y, y1_end], color=color,
                           linewidth=2, linestyle="--", label=f"Phase {pm.phase} Linear", alpha=0.8)

                    # Segment 2: from day_range[0] to end_day
                    x2_start = day_range[0]
                    x2_end = pm.end_day
                    y2_start = y1_end  # Continue from where segment 1 ended
                    ax.plot([x2_start, x2_end], [y2_start, end_y], color=color,
                           linewidth=2, linestyle="--", alpha=0.8)
                else:
                    # Normal phase within range - draw line from turning point to turning point
                    ax.plot([pm.start_day, pm.end_day], [start_y, end_y], color=color,
                           linewidth=2, linestyle="--", label=f"Phase {pm.phase} Linear", alpha=0.8)

        # Add phase model summary text
        phase_text = "Phase Slopes:\n"
        for pm in phase_models:
            sig = "***" if pm.p_value < 0.001 else "**" if pm.p_value < 0.01 else "*" if pm.p_value < 0.05 else ""
            phase_text += f"  Phase {pm.phase} (Day {pm.start_day:.0f} to {pm.end_day:.0f}): "
            phase_text += f"{pm.slope:.3f} ± {pm.slope_se:.3f} {sig}\n"

        ax.text(
            0.98, 0.02, phase_text,
            transform=ax.transAxes,
            fontsize=9,
            verticalalignment="bottom",
            horizontalalignment="right",
            family="monospace",
            bbox=dict(boxstyle="round,pad=0.5", facecolor="white", edgecolor="gray", alpha=0.9)
        )

    # Formatting
    ax.set_xlabel(xlabel, fontsize=12)
    ax.set_ylabel(ylabel, fontsize=12)
    ax.set_title(title, fontsize=14, fontweight="bold")

    ax.set_xlim(day_range[0] - 0.5, day_range[1] + 0.5)
    ax.set_xticks(range(day_range[0], day_range[1] + 1, 2))

    ax.legend(loc="upper right")
    ax.grid(True, alpha=0.3)

    plt.tight_layout()

    if save_path:
        plt.savefig(save_path, dpi=300, bbox_inches="tight")

    return fig


def plot_confounder_effects(
    confounder_results: List,
    title: str = "Confounder Effects on Cycle-Outcome Relationship",
    figsize: Tuple[int, int] = (10, 6),
    save_path: Optional[str] = None
) -> plt.Figure:
    """
    Plot forest plot of confounder effects.

    Parameters
    ----------
    confounder_results : List[ConfounderResult]
        Results from analyze_confounders()
    title : str
        Plot title
    figsize : Tuple[int, int]
        Figure size
    save_path : str, optional
        Path to save figure

    Returns
    -------
    plt.Figure
        Matplotlib figure object
    """
    if not confounder_results:
        fig, ax = plt.subplots(figsize=figsize)
        ax.text(0.5, 0.5, "No confounder results to display",
                ha="center", va="center", transform=ax.transAxes)
        return fig

    fig, ax = plt.subplots(figsize=figsize)

    confounders = [r.confounder for r in confounder_results]
    coefs = [r.coefficient for r in confounder_results]
    ci_lower = [r.ci_lower for r in confounder_results]
    ci_upper = [r.ci_upper for r in confounder_results]
    p_values = [r.p_value for r in confounder_results]

    y_pos = np.arange(len(confounders))

    # Colors based on significance
    colors = ["red" if p < 0.05 else "gray" for p in p_values]

    # Plot
    ax.errorbar(
        coefs, y_pos,
        xerr=[np.array(coefs) - np.array(ci_lower), np.array(ci_upper) - np.array(coefs)],
        fmt="o",
        capsize=5,
        capthick=2,
        color="steelblue",
        ecolor="steelblue",
        markersize=8
    )

    # Add significance markers
    for i, (coef, p) in enumerate(zip(coefs, p_values)):
        sig = "***" if p < 0.001 else "**" if p < 0.01 else "*" if p < 0.05 else ""
        ax.annotate(sig, (coef, i), xytext=(5, 0), textcoords="offset points",
                   fontsize=12, fontweight="bold", color="red")

    ax.axvline(x=0, color="gray", linestyle="--", alpha=0.5)
    ax.set_yticks(y_pos)
    ax.set_yticklabels(confounders)
    ax.set_xlabel("Coefficient (effect on normalized outcome)")
    ax.set_title(title, fontsize=14, fontweight="bold")
    ax.grid(True, alpha=0.3, axis="x")

    plt.tight_layout()

    if save_path:
        plt.savefig(save_path, dpi=300, bbox_inches="tight")

    return fig

File: r-package/mcanalysis/R/visualization.R
#' Plot Cycle Effect
#'
#' Plot GAM results with confidence intervals and optional phase information.
#'
#' @param gam_result Result from fit_gam()
#' @param turning_points Optional turning points to mark
#' @param phase_models Optional phase model results
#' @param raw_data Optional processed data for showing daily means
#' @param title Plot title
#' @param xlabel X-axis label
#' @param ylabel Y-axis label
#' @param show_significance Whether to show significance annotation (default TRUE)
#' @param show_phases Whether to shade phases (default TRUE)
#' @param show_linear_models Whether to show linear phase lines (default TRUE)
#' @param show_turning_points Whether to mark turning points (default TRUE)
#' @param show_ci Whether to show confidence interval (default TRUE)
#' @param show_period_line Whether to show vertical line at day 0 (default TRUE)
#' @param show_raw_data Whether to show daily means line (default FALSE)
#' @param day_range Range of days to display (default c(-14, 13))
#' @param gam_color Color of GAM curve (default "steelblue")
#' @param gam_linewidth Line width of GAM curve (default 1)
#' @param ci_alpha Transparency of CI (default 0.3)
#' @param linear_colors Colors for linear phase models (default c("#E74C3C", "#27AE60"))
#' @param raw_data_alpha Transparency of daily means line (default 0.5)
#' @param raw_data_color Color of daily means line (default "gray50")
#'
#' @return ggplot2 object
#' @export
#'
#' @import ggplot2
plot_cycle_effect <- function(gam_result,
                               turning_points = NULL,
                               phase_models = NULL,
                               raw_data = NULL,
                               title = "Menstrual Cycle Effect on Outcome",
                               xlabel = "Cycle Day (Day 0 = Period Start)",
                               ylabel = NULL,
                               show_significance = TRUE,
                               show_phases = TRUE,
                               show_linear_models = TRUE,
                               show_turning_points = TRUE,
                               show_ci = TRUE,
                               show_period_line = TRUE,
                               show_raw_data = FALSE,
                               day_range = c(-14, 13),
                               gam_color = "steelblue",
                               gam_linewidth = 1,
                               ci_alpha = 0.3,
                               linear_colors = c("#E74C3C", "#27AE60"),
                               raw_data_alpha = 0.5,
                               raw_data_color = "gray50") {

  pred_df <- gam_result$predictions
  is_binary <- isTRUE(gam_result$is_binary)

  # Filter to day range
  plot_data <- pred_df[pred_df$cycle_day >= day_range[1] &
                        pred_df$cycle_day <= day_range[2], ]

  if (is_binary) {
    # For binary outcomes: use log odds
    if (is.null(ylabel)) ylabel <- "Log Odds"

    # Center log odds around mean for plotting
    mean_logit <- mean(plot_data$log_odds)
    plot_data$predicted_pct <- plot_data$log_odds - mean_logit

    # CI from bootstrap on log odds scale (if available) or from probability
    if ("or_ci_lower" %in% names(plot_data)) {
      plot_data$ci_lower_pct <- log(plot_data$or_ci_lower)
      plot_data$ci_upper_pct <- log(plot_data$or_ci_upper)
    } else {
      # Approximate CI on log odds from probability CI
      eps <- 1e-6
      plot_data$ci_lower_pct <- log(pmax(plot_data$ci_lower, eps) / pmax(1 - plot_data$ci_lower, eps)) - mean_logit
      plot_data$ci_upper_pct <- log(pmax(plot_data$ci_upper, eps) / pmax(1 - plot_data$ci_upper, eps)) - mean_logit
    }
  } else {
    # For continuous outcomes: percentage change from 100
    if (is.null(ylabel)) ylabel <- "% Change from Personal Average"

    plot_data$predicted_pct <- plot_data$predicted - 100
    plot_data$ci_lower_pct <- plot_data$ci_lower - 100
    plot_data$ci_upper_pct <- plot_data$ci_upper - 100
  }

  # Calculate y-axis range
  y_min <- min(plot_data$ci_lower_pct, na.rm = TRUE)
  y_max <- max(plot_data$ci_upper_pct, na.rm = TRUE)
  y_range <- y_max - y_min

  # Base plot
  p <- ggplot2::ggplot(plot_data, ggplot2::aes(x = cycle_day))

  # Add daily means line if requested
  if (show_raw_data && !is.null(raw_data)) {
    raw_filtered <- raw_data[raw_data$cycle_day >= day_range[1] &
                              raw_data$cycle_day <= day_range[2] &
                              !raw_data$is_dummy, ]
    daily_means <- stats::aggregate(a ~ cycle_day, data = raw_filtered, FUN = mean)
    daily_means$a_pct <- daily_means$a - 100

    p <- p +
      ggplot2::geom_line(
        data = daily_means,
        ggplot2::aes(x = cycle_day, y = a_pct),
        color = raw_data_color,
        alpha = raw_data_alpha,
        linewidth = 0.8
      )
  }

  # Confidence interval (no legend mapping)
  if (show_ci) {
    p <- p +
      ggplot2::geom_ribbon(
        ggplot2::aes(ymin = ci_lower_pct, ymax = ci_upper_pct),
        alpha = ci_alpha,
        fill = gam_color
      )
  }

  # GAM curve (no legend mapping)
  p <- p +
    ggplot2::geom_line(
      ggplot2::aes(y = predicted_pct),
      color = gam_color,
      linewidth = gam_linewidth
    )

  # Reference line at 0 (mean)
  p <- p +
    ggplot2::geom_hline(yintercept = 0, linetype = "dashed", color = "gray50", linewidth = 0.6)

  # Period start line
  if (show_period_line) {
    p <- p +
      ggplot2::geom_vline(xintercept = 0, linetype = "dotted", color = "red", linewidth = 0.8)
  }

  # Add phase shading
  if (show_phases && !is.null(phase_models) && nrow(phase_models) > 0) {
    phase_colors <- c("#E8F4F8", "#FFF3CD")
    for (i in seq_len(nrow(phase_models))) {
      pm <- phase_models[i, ]
      fill_color <- phase_colors[(i - 1) %% length(phase_colors) + 1]

      if (pm$end_day <= pm$start_day) {
        p <- p +
          ggplot2::annotate("rect",
            xmin = pm$start_day, xmax = day_range[2],
            ymin = -Inf, ymax = Inf,
            alpha = 0.1, fill = fill_color) +
          ggplot2::annotate("rect",
            xmin = day_range[1], xmax = pm$end_day,
            ymin = -Inf, ymax = Inf,
            alpha = 0.1, fill = fill_color)
      } else {
        p <- p +
          ggplot2::annotate("rect",
            xmin = pm$start_day, xmax = pm$end_day,
            ymin = -Inf, ymax = Inf,
            alpha = 0.1, fill = fill_color)
      }
    }
  }

  # Add linear model lines
  if (show_linear_models && !is.null(phase_models) && nrow(phase_models) > 0 && !is.null(turning_points)) {
    for (i in seq_len(nrow(phase_models))) {
      pm <- phase_models[i, ]
      line_color <- linear_colors[(i - 1) %% length(linear_colors) + 1]

      start_idx <- which.min(abs(plot_data$cycle_day - pm$start_day))
      end_idx <- which.min(abs(plot_data$cycle_day - pm$end_day))
      start_y <- plot_data$predicted_pct[start_idx]
      end_y <- plot_data$predicted_pct[end_idx]

      if (pm$end_day <= pm$start_day) {
        total_span <- (day_range[2] - pm$start_day) + (pm$end_day - day_range[1])
        frac1 <- (day_range[2] - pm$start_day) / total_span
        y_at_edge <- start_y + frac1 * (end_y - start_y)

        p <- p +
          ggplot2::annotate("segment",
            x = pm$start_day, xend = day_range[2],
            y = start_y, yend = y_at_edge,
            color = line_color, linetype = "dashed", linewidth = 0.8, alpha = 0.8
          ) +
          ggplot2::annotate("segment",
            x = day_range[1], xend = pm$end_day,
            y = y_at_edge, yend = end_y,
            color = line_color, linetype = "dashed", linewidth = 0.8, alpha = 0.8
          )
      } else {
        p <- p +
          ggplot2::annotate("segment",
            x = pm$start_day, xend = pm$end_day,
            y = start_y, yend = end_y,
            color = line_color, linetype = "dashed", linewidth = 0.8, alpha = 0.8
          )
      }
    }
  }

  # Define legend position for collision detection
  legend_x_min <- day_range[2] - 6
  legend_x_max <- day_range[2]
  legend_y_max <- y_max
  legend_y_min <- y_max - y_range * 0.18

  # Add turning points with collision detection
  if (show_turning_points && !is.null(turning_points)) {
    for (i in seq_along(turning_points)) {
      tp <- turning_points[i]
      if (tp >= day_range[1] && tp <= day_range[2]) {
        closest_idx <- which.min(abs(plot_data$cycle_day - tp))
        y_val <- plot_data$predicted_pct[closest_idx]

        # Check if label would collide with legend (top right area)
        in_legend_x <- tp > legend_x_min
        in_legend_y_above <- y_val > legend_y_min

        # Determine label position - avoid legend collision and stay in bounds
        if (in_legend_x && in_legend_y_above) {
          # Would collide with legend - put label below point
          label_vjust <- 1.5
        } else if (y_val > (y_max - y_range * 0.12)) {
          # Near top but not in legend area - put label below
          label_vjust <- 1.5
        } else {
          # Safe to put label above
          label_vjust <- -0.5
        }

        p <- p +
          ggplot2::geom_vline(xintercept = tp, linetype = "dashed",
                              color = "orange", alpha = 0.6) +
          ggplot2::geom_point(
            data = data.frame(x = tp, y = y_val),
            ggplot2::aes(x = x, y = y),
            color = "orange", size = 4
          ) +
          ggplot2::annotate(
            "label",
            x = tp, y = y_val,
            label = paste0("TP ", i, " (Day ", round(tp, 1), ")"),
            size = 3.5,
            vjust = label_vjust,
            label.padding = ggplot2::unit(0.25, "lines"),
            label.r = ggplot2::unit(0.2, "lines"),
            fill = "white",
            color = "gray30"
          )
      }
    }
  }

  # Add significance annotation as subtitle
  if (show_significance) {
    p_val <- gam_result$p_value
    sig_stars <- if (p_val < 0.001) "***" else if (p_val < 0.01) "**" else if (p_val < 0.05) "*" else "(ns)"

    if (is_binary) {
      # Binary outcome: show pseudo R² and peak-to-trough OR
      or_text <- if (!is.null(gam_result$peak_trough_or)) {
        sprintf("  |  Peak-Trough OR: %.2f", gam_result$peak_trough_or)
      } else ""
      sig_text <- paste0("Cycle Effect p-value: ", format(p_val, scientific = TRUE, digits = 2), " ", sig_stars,
                         "  |  Pseudo R²: ", round(gam_result$deviance_explained * 100, 1), "%",
                         or_text)
    } else {
      sig_text <- paste0("Cycle Effect p-value: ", format(p_val, scientific = TRUE, digits = 2), " ", sig_stars,
                         "  |  Deviance Explained: ", round(gam_result$deviance_explained * 100, 1), "%")
    }

    p <- p + ggplot2::labs(subtitle = sig_text)
  }

  # === LEGEND (top right corner) ===
  ci_symbol <- "\u2588"
  gam_symbol <- "\u2014"
  legend_text <- paste0(ci_symbol, "  95% CI\n", gam_symbol, "  GAM Fit")

  p <- p +
    ggplot2::annotate(
      "label",
      x = day_range[2],
      y = y_max,
      label = legend_text,
      hjust = 1, vjust = 1,
      size = 4,
      fontface = "bold",
      color = gam_color,
      fill = "white",
      label.padding = ggplot2::unit(0.4, "lines"),
      label.r = ggplot2::unit(0.3, "lines")
    )

  # === PHASE SLOPES (bottom right corner) ===
  if (!is.null(phase_models) && nrow(phase_models) > 0) {
    phase_text <- "Phase Slopes:\n"
    for (i in seq_len(nrow(phase_models))) {
      pm <- phase_models[i, ]
      sig <- if (pm$p_value < 0.001) "***" else if (pm$p_value < 0.01) "**" else if (pm$p_value < 0.05) "*" else ""
      phase_text <- paste0(phase_text,
        sprintf("Phase %d (Day %.0f to %.0f): %.3f +/- %.3f %s\n",
                pm$phase, pm$start_day, pm$end_day,
                pm$slope, pm$slope_se, sig))
    }

    p <- p +
      ggplot2::annotate(
        "label",
        x = day_range[2],
        y = y_min,
        label = phase_text,
        hjust = 1, vjust = 0,
        size = 3.5,
        fontface = "bold",
        color = "gray30",
        fill = "white",
        label.padding = ggplot2::unit(0.4, "lines"),
        label.r = ggplot2::unit(0.3, "lines")
      )
  }

  # === WATERMARK (bottom left corner) ===
  p <- p +
    ggplot2::annotate(
      "text",
      x = day_range[1],
      y = y_min,
      label = "Produced by MCAnalysis",
      hjust = 0, vjust = 0,
      size = 3.5,
      fontface = "italic",
      color = "gray50"
    )

  # Formatting with consistent font
  p <- p +
    ggplot2::scale_x_continuous(breaks = seq(day_range[1], day_range[2], by = 2)) +
    ggplot2::labs(
      title = title,
      x = xlabel,
      y = ylabel
    ) +
    ggplot2::theme_minimal(base_size = 12) +
    ggplot2::theme(
      text = ggplot2::element_text(size = 12),
      plot.title = ggplot2::element_text(face = "bold", size = 14),
      plot.subtitle = ggplot2::element_text(size = 11),
      axis.title = ggplot2::element_text(size = 12),
      axis.text = ggplot2::element_text(size = 11),
      panel.grid.minor = ggplot2::element_blank(),
      legend.position = "none"
    )

  return(p)
}


#' Plot Confounder Effects
#'
#' Plot forest plot of confounder effects.
#'
#' @param confounder_results Results from analyze_confounders()
#' @param title Plot title
#'
#' @return ggplot2 object
#' @export
#'
#' @import ggplot2
plot_confounder_effects <- function(confounder_results,
                                     title = "Confounder Effects on Cycle-Outcome Relationship") {

  if (is.null(confounder_results) || nrow(confounder_results) == 0) {
    message("No confounder results to display")
    return(NULL)
  }

  # Add significance indicator
  confounder_results$significant <- confounder_results$p_value < 0.05

  p <- ggplot2::ggplot(
    confounder_results,
    ggplot2::aes(x = coefficient, y = stats::reorder(confounder, coefficient))
  ) +
    ggplot2::geom_vline(xintercept = 0, linetype = "dashed", color = "gray50") +
    ggplot2::geom_errorbarh(
      ggplot2::aes(xmin = ci_lower, xmax = ci_upper),
      height = 0.2,
      color = "steelblue"
    ) +
    ggplot2::geom_point(
      ggplot2::aes(color = significant),
      size = 3
    ) +
    ggplot2::scale_color_manual(
      values = c("TRUE" = "red", "FALSE" = "steelblue"),
      guide = "none"
    ) +
    ggplot2::labs(
      title = title,
      x = "Coefficient (effect on normalized outcome)",
      y = ""
    ) +
    ggplot2::theme_minimal(base_size = 12) +
    ggplot2::theme(
      text = ggplot2::element_text(size = 12),
      plot.title = ggplot2::element_text(face = "bold", size = 14),
      axis.title = ggplot2::element_text(size = 12),
      axis.text = ggplot2::element_text(size = 11)
    )

  return(p)
}
